page.go

```diff
@@ -79,40 +77,21 @@ func (p *page) writeToNoChecksum(w io.Writer) error {
 		nextPagePosition = math.MaxUint64
 	}
 
-	// write pageStatus, transactionNumber, nextPage, and payload size
-	buf := make([]byte, 8+8+8+8)
-	binary.LittleEndian.PutUint64(buf[0:8], p.pageStatus)
-	binary.LittleEndian.PutUint64(buf[8:16], p.transactionNumber)
-	binary.LittleEndian.PutUint64(buf[16:24], nextPagePosition)
-	binary.LittleEndian.PutUint64(buf[24:32], uint64(len(p.payload)))
-	if _, err := w.Write(buf); err != nil {
-		return err
-	}
-
-	// write payload
-	if _, err := w.Write(p.payload); err != nil {
-		return err
-	}
-
-	return nil
-}
-
-func (p *page) writeTo(w io.Writer) error {
-	// write checksum
-	if _, err := w.Write(p.transactionChecksum[:]); err != nil {
-		return err
+	// if buf has enough capacity to hold p, use it; otherwise allocate
+	var b []byte
+	if rest := buf[len(buf):]; cap(rest) >= p.size() {
+		b = rest[:p.size()]
```

Comment:
this is valid?

Reply:
Seems like it. I tested it on the [playground](https://play.golang.org/p/p2RRw1R3Y3)

Reply:
it looks a little strange ("rest is already at the end of the slice, how could you reslice it to be larger?") but it works because you can reslice up to the `cap` of the slice, not just the `len`.

It might look less scary as:
```go
if cap(buf) - len(buf) >= p.size() {
    b = buf[len(buf) : len(buf)+p.size()]
}
```